server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -240,42 +379,58 @@ private ByteSizeValue getFreeBytesThreshold(ByteSizeValue total, RelativeByteSiz
     }
 
     public ByteSizeValue getFreeBytesThresholdLowStage(ByteSizeValue total) {
-        return getFreeBytesThreshold(total, lowStageWatermark, ByteSizeValue.MINUS_ONE);
+        return getFreeBytesThreshold(total, lowStageWatermark, lowStageMaxHeadroom);
     }
 
     public ByteSizeValue getFreeBytesThresholdHighStage(ByteSizeValue total) {
-        return getFreeBytesThreshold(total, highStageWatermark, ByteSizeValue.MINUS_ONE);
+        return getFreeBytesThreshold(total, highStageWatermark, highStageMaxHeadroom);
     }
 
     public ByteSizeValue getFreeBytesThresholdFloodStage(ByteSizeValue total) {
-        return getFreeBytesThreshold(total, floodStageWatermark, ByteSizeValue.MINUS_ONE);
+        return getFreeBytesThreshold(total, floodStageWatermark, floodStageMaxHeadroom);
     }
 
     public ByteSizeValue getFreeBytesThresholdFrozenFloodStage(ByteSizeValue total) {
         return getFreeBytesThreshold(total, frozenFloodStageWatermark, frozenFloodStageMaxHeadroom);
     }
 
-    public ByteSizeValue getMinimumTotalSizeForBelowLowWatermark(ByteSizeValue used) {
+    private ByteSizeValue getMinimumTotalSizeForBelowWatermark(
+        ByteSizeValue used,
+        RelativeByteSizeValue watermark,
+        ByteSizeValue maxHeadroom
+    ) {
         // If watermark is absolute, simply return total disk = used disk + free disk, where free disk bytes is the watermark value.
-        if (lowStageWatermark.isAbsolute()) {
-            return ByteSizeValue.ofBytes(lowStageWatermark.getAbsolute().getBytes() + used.getBytes());
+        if (watermark.isAbsolute()) {
+            return ByteSizeValue.ofBytes(watermark.getAbsolute().getBytes() + used.getBytes());
         }
 
         // If watermark is percentage/ratio, calculate the total needed disk space.
-        double percentThreshold = lowStageWatermark.getRatio().getAsPercent();
+        // This may not be the minimum, due to the possible max headroom value which can cap the free disk space required.
+        double percentThreshold = watermark.getRatio().getAsPercent();
         if (percentThreshold >= 0.0 && percentThreshold < 100.0) {
-            // Use percentage instead of ratio, and multiple bytes with 100, to make division with double more accurate (issue #88791).
+            // Use percentage instead of ratio, and multiply bytes with 100, to make division with double more accurate (issue #88791).
             ByteSizeValue totalBytes = ByteSizeValue.ofBytes((long) Math.ceil((100 * used.getBytes()) / percentThreshold));
-            return totalBytes;
+
+            // Now calculate the minimum free bytes, taking into account the possible max headroom value as well.
+            ByteSizeValue minimumFreeBytes = ByteSizeValue.ofBytes(
+                totalBytes.getBytes() - watermark.calculateValueWithFloor(totalBytes, maxHeadroom).getBytes()
+            );
```

Comment:
Hi @henningandersen , I could do that but I would need a bit of input/guidance on the helper functions:
* What should min() do if one of the arguments is -1? ignore it I guess and return the other. What in the case both of them are -1? Throw?
* I could implement add() as I did (adding bytes and returning a new ByteSizeValue) but then unfortunately any stored unit would be lost (and effectively they would be bytes). So e.g. doing add(1k, 1m) would return a 1049600b ByteSizeValue.
I could also do these in a separate PR since they are orthogonal.

Reply:
> What should min() do if one of the arguments is -1?

I think we should throw in case any of them are -1.

> add(1k, 1m) would return a 1049600b ByteSizeValue.

I think that will work fine, we should probably add a comment that the resulting unit is not defined by the function.

Reply:
Thank you @henningandersen , after revisiting the code, please note that we do not need to do a "min(a,b)" operation at all. Rather the "minimum" I mentioned in the code referred to the possibility of a max headroom further decreasing the total size required. For this reason, I changed the variable name from "minimumFreeBytes" to "cappedFreeBytes" and also enriched the comments to make it more understandable. I did introduce though your suggested ByteSizeValue.add() and .subtract() functions which I now use in this code and makes the code more readable. I added also relevant tests for these two new functions. Please feel free to review and unresolve this conversation if you feel it's not handled sufficiently.

Reply:
We discussed this again and Iraklis will try to follow the suggestion above.